d2l-table-wrapper.html

```diff
@@ -143,14 +145,110 @@
 			}
 		</style>
 		<d2l-scroll-wrapper show-actions>
-			<slot></slot>
+			<slot id="slot"></slot>
 		</d2l-scroll-wrapper>
 	</template>
 	<script>
 		Polymer({
 			is: 'd2l-table-wrapper',
+			behaviors: [
+				D2L.PolymerBehaviors.Table.LocalObserverBehavior
+			],
+			listeners: {
+				'd2l-table-local-observer': '_handleLocalObserver'
+			},
+			__applyInQueue: false,
+			__tableObserver: null,
+			attached: function() {
+				Polymer.RenderStatus.afterNextRender(this, function() {
+					this._registerTableObserver();
+					this._applyClassNames();
+				}.bind(this));
+			},
+			detached: function() {
+				if (this.__tableObserver !== null) {
+					this.__tableObserver.disconnect();
+					this.__tableObserver = null;
+				}
+			},
 			notifyResize: function() {
 				this.$$('d2l-scroll-wrapper').notifyResize();
+			},
+			_applyClassNames: function() {
+
+				// don't allow a bunch of calls to queue up
+				if (this.__applyInQueue) return;
+				this.__applyInQueue = true;
+
+				fastdom.mutate(function() {
+
+					this.__applyInQueue = false;
+
+					var table = this.queryEffectiveChildren('.d2l-table');
+					if (!table) return;
+
+					var firstRow = null;
+					var lastRow = null;
+
+					for (var i = 0; i < table.rows.length; i++) {
+						var row = table.rows[i];
+						row.classList.remove('d2l-table-row-first');
+						row.classList.remove('d2l-table-row-last');
+						if (row.offsetParent !== null) {
+							if (firstRow === null) {
```

Comment:
Nitpick - these null checks can probably be simplified.  Not a big deal though... code is clean and easy to follow.

Reply:
Oh... this `row.offsetParent` would be considered a read, but the code is expecting to be doing mutations.  This visibility measurement could perhaps be done elsewhere.

Reply:
Since we're using fast-dom, and we're in a mutation phase, this might not be a big concern, but I will mention it just in case... it's possible that we are removing `d2l-table-row-first` and `d2l-table-row-last` unnecessarily.  i.e. those classes may be reapplied to the same nodes below at line 219.  The browser may unnecessarily think layout is dirty.  In theory this could should be able to run in certain scenarios without making the layout dirty.

Reply:
Those are both excellent points. I'm going to have another look at this and see if I can't resolve both of them.